polkadot/runtime/parachains/src/disputes.rs

```diff
@@ -1216,12 +1235,13 @@ impl<T: Config> Pallet<T> {
 		let revert_to = included_in - One::one();
 
 		Included::<T>::insert(&session, &candidate_hash, revert_to);
+	}
 
-		if let Some(state) = Disputes::<T>::get(&session, candidate_hash) {
-			if has_supermajority_against(&state) {
-				Self::revert_and_freeze(revert_to);
-			}
-		}
+	pub(crate) fn disputes_concluded_invalid(session: SessionIndex) -> BTreeSet<CandidateHash> {
+		Disputes::<T>::iter_prefix(session)
```

Comment:
Doesn't this become quite expensive with a lot of disputes in a single session, increasing the block import time ?

Reply:
wouldn't it be enough if we were to check the candidates concluded against in the last `allowed_ancestry_len` blocks?

Reply:
> Doesn't this become quite expensive with a lot of disputes in a single session, increasing the block import time ?

That's a fair point. I think we can change it to be lazily computed based on the included candidates, will take a look.

> wouldn't it be enough if we were to check the candidates concluded against that are anchored in the last allowed_ancestry_len blocks?

It would be, but the disputes are stored in a double map based on session index (for pruning efficiency). We could change that to relay parent block number, but that might complicate other things. I would try to explore a different avenue suggested above first.